### PR TITLE
chore(e2e-test) Perform operations using dynamic go-client approach

### PR DESCRIPTION
**Pull request by oumkale**
This PR has been using go-client approach for e2e test operations:

 - Currently, there are some of the commands which are in kubectl apply -f to create some yaml file. 
 - This commands are getting changed with a go client approach. 
 - Go structure for Experiment and Engine

Operations using common structure to Install Chaos resources :
```
//ChaosExperiment is go struct for Experiment
type ChaosExperiment struct {
	APIVersion  string `json:"apiVersion,omitempty"`
	Description struct {
		Message string `json:"message,omitempty"`
	} `json:"description,omitempty"`
	Kind     string `json:"kind,omitempty"`
	Metadata struct {
		Name   string `json:"name,omitempty"`
		Labels struct {
			Name                     string `json:"name,omitempty"`
			AppKubernetesIoPartOf    string `json:"app.kubernetes.io/part-of,omitempty"`
			AppKubernetesIoComponent string `json:"app.kubernetes.io/component,omitempty"`
			AppKubernetesIoVersion   string `json:"app.kubernetes.io/version,omitempty"`
		} `json:"labels,omitempty"`
	} `json:"metadata,omitempty"`
	Spec struct {
		Definition struct {
			Scope       string `json:"scope,omitempty"`
			Permissions []struct {
				APIGroups []string `json:"apiGroups,omitempty"`
				Resources []string `json:"resources,omitempty"`
				Verbs     []string `json:"verbs,omitempty"`
			} `json:"permissions,omitempty"`
			Image           string   `json:"image,omitempty"`
			ImagePullPolicy string   `json:"imagePullPolicy,omitempty"`
			Args            []string `json:"args,omitempty"`
			Command         []string `json:"command,omitempty"`
			Env             []struct {
				Name  string `json:"name"`
				Value string `json:"value"`
			} `json:"env,omitempty"`
			Labels struct {
				Name                     string `json:"name,omitempty"`
				AppKubernetesIoPartOf    string `json:"app.kubernetes.io/part-of,omitempty"`
				AppKubernetesIoComponent string `json:"app.kubernetes.io/component,omitempty"`
				AppKubernetesIoVersion   string `json:"app.kubernetes.io/version,omitempty"`
			} `json:"labels,omitempty"`
		} `json:"definition,omitempty"`
	} `json:"spec,omitempty"`
}
```

```
//ChaosEngine is go struct for Engine
type ChaosEngine struct {
	APIVersion string `json:"apiVersion,omitempty"`
	Kind       string `json:"kind,omitempty"`
	Metadata   struct {
		Name      string `json:"name,omitempty"`
		Namespace string `json:"namespace,omitempty"`
	} `json:"metadata,omitempty"`
	Spec struct {
		AnnotationCheck string `json:"annotationCheck,omitempty"`
		EngineState     string `json:"engineState,omitempty"`
		Appinfo         struct {
			Appns    string `json:"appns,omitempty"`
			Applabel string `json:"applabel,omitempty"`
			Appkind  string `json:"appkind,omitempty"`
		} `json:"appinfo"`
		ChaosServiceAccount string `json:"chaosServiceAccount,omitempty"`
		JobCleanUpPolicy    string `json:"jobCleanUpPolicy,omitempty"`
		Components          struct {
			Runner struct {
				ImagePullPolicy string `json:"imagePullPolicy,omitempty"`
			} `json:"runner,omitempty"`
		} `json:"components,omitempty"`
		Experiments []struct {
			Name string `json:"name,omitempty"`
			Spec struct {
				Components struct {
					NodeSelector struct {
						KubernetesIoHostname string `json:"kubernetes.io/hostname,omitempty"`
					} `json:"nodeSelector,omitempty"`
					Env []struct {
						Name  string `json:"name"`
						Value string `json:"value"`
					} `json:"env,omitempty"`
				} `json:"components,omitempty"`
			} `json:"spec,omitempty"`
		} `json:"experiments,omitempty"`
	} `json:"spec,omitempty"`
}
```

#### Issue
- fixes: https://github.com/litmuschaos/litmus/issues/1487

### CODE REVIEW

File: pkg/install.go [hunk 1]
```diff
@@ -1,14 +1,92 @@
...
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/restmapper"
 )
 
+//CreateChaosResource Create the ChaosResource
```
Comment (uditgaurav) [inline pkg/install.go:28]:
```suggestion
//CreateChaosResource creates litmus components with given inputs
```

File: pkg/install.go [hunk 2]
```diff
@@ -37,55 +115,79 @@ func InstallGoRbac(testsDetails *types.TestDetails, rbacNamespace string) error
...
+		return errors.Errorf("Fail to fetch the rbac file, due to %v", err)
 	}
-	// Modify the spec of experiemnt file
-	if err = EditFile("/tmp/"+testsDetails.ExperimentName+"-exp.yaml", "image: \"litmuschaos/go-runner:latest\"", "image: "+testsDetails.GoExperimentImage); err != nil {
-		return errors.Errorf("Fail to Update the experiment file, due to %v", err)
+
+	// ReadAll reads from response until an error or EOF and returns the data it read.
+	fileInput, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		log.Fatalf("Fail to read data from response: %v", err)
```
Comment (uditgaurav) [inline pkg/install.go:132]:
```suggestion
		log.Errorf("Fail to read data from response: %v", err)
```

File: pkg/install.go [hunk 3]
```diff
@@ -37,55 +115,79 @@ func InstallGoRbac(testsDetails *types.TestDetails, rbacNamespace string) error
...
 	}
-	// Modify experiment imagePullPolicy
-	if err = EditFile("/tmp/"+testsDetails.ExperimentName+"-exp.yaml", "imagePullPolicy: Always", "imagePullPolicy: "+testsDetails.ExperimentImagePullPolicy); err != nil {
-		log.Info("Field imagePullPolicy not defined")
+
+	// Unmarshal decodes
+	err = yamlChe.Unmarshal([]byte(fileInput), &chaosExperiment)
+	if err != nil {
+		log.Fatalf("error when unmarshelling: %v", err)
```
Comment (uditgaurav) [inline pkg/install.go:138]:
```suggestion
		log.Errorf("error when unmarshelling: %v", err)
```

File: pkg/install.go [hunk 4]
```diff
@@ -95,91 +197,92 @@ func InstallGoChaosExperiment(testsDetails *types.TestDetails, experimentNamespa
...
-	}
-	if err = EditFile("/tmp/"+testsDetails.ExperimentName+"-ce.yaml", "name: "+testsDetails.ExperimentName+"", "name: "+testsDetails.NewExperimentName+""); err != nil {
-		return errors.Errorf("Fail to Update the engine file, due to %v", err)
-	}
-	if err = EditFile("/tmp/"+testsDetails.ExperimentName+"-ce.yaml", "jobCleanUpPolicy: 'delete'", "jobCleanUpPolicy: "+testsDetails.JobCleanUpPolicy+""); err != nil {
-		return errors.Errorf("Fail to Update the engine file, due to %v", err)
-	}
-	if err = AddAfterMatch("/tmp/"+testsDetails.ExperimentName+"-ce.yaml", "engineState: 'active'", "  annotationCheck: '"+testsDetails.AnnotationCheck+"'"); err != nil {
-		return errors.Errorf("Fail to Update the engine file, err: %v", err)
+		chaosEngine.Spec.Appinfo.Appns = testsDetails.AppNS
```
[the remark this answers is not in the document]
Comment (uditgaurav) [reply pkg/install.go:235]: Do we need this condition `if testsDetails.AppNS != "" && testsDetails.AppLabel != "" {` ?

File: pkg/install.go [hunk 5]
```diff
@@ -95,91 +197,92 @@ func InstallGoChaosExperiment(testsDetails *types.TestDetails, experimentNamespa
...
 	if testsDetails.ExperimentName == "pod-cpu-hog" {
-		if err = AddAfterMatch("/tmp/"+testsDetails.ExperimentName+"-ce.yaml", "value: '60'", "\n            - name: CHAOS_KILL_COMMAND\n              value: "+testsDetails.CPUKillCommand+""); err != nil {
-			return errors.Errorf("Failed to add the chaos kill command,due to %v", err)
-		}
-	} else if testsDetails.ExperimentName == "pod-memory-hog" {
-		if err = AddAfterMatch("/tmp/"+testsDetails.ExperimentName+"-ce.yaml", "value: '60'", "\n            - name: CHAOS_KILL_COMMAND\n              value: "+testsDetails.MemoryKillCommand+""); err != nil {
-			return errors.Errorf("Failed to add the chaos kill command,due to %v", err)
-		}
+		environments["CHAOS_KILL_COMMAND"] = testsDetails.CPUKillCommand
+	} else {
```
Comment (uditgaurav) [inline pkg/install.go:257]: elseif `testsDetails.ExperimentName == "pod-memory-hog" {`

File: pkg/install.go [hunk 6]
```diff
@@ -95,91 +197,92 @@ func InstallGoChaosExperiment(testsDetails *types.TestDetails, experimentNamespa
...
+	// update all the value corresponding to keys from the ENV's in Engine
+	for key, value := range chaosEngine.Spec.Experiments[0].Spec.Components.Env {
+		_, ok := environments[value.Name]
+		if ok {
+			chaosEngine.Spec.Experiments[0].Spec.Components.Env[key].Value = environments[value.Name]
 		}
 	}
+
+	// Marshal serializes the value provided into a YAML document.
+	fileData, _ := json.Marshal(chaosEngine)
```
Comment (uditgaurav) [inline pkg/install.go:281]: please check the err here.

**Review by Jonsy13 — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Comment by uditgaurav** [comment]
The successful pipeline runs:

- Component-Pipeline: https://github.com/litmuschaos/litmus-e2e/actions/runs/835703676
- Pod-Level-Pipeline: https://github.com/litmuschaos/litmus-e2e/actions/runs/835704207

**Pull request merged**